src/databases/NASTRAN/avtNASTRANFileFormat.C

```diff
@@ -187,90 +187,75 @@ avtNASTRANFileFormat::ActivateTimestep()
 //    Ok, I 'fixed' this funky logic again. The above 'fix' caused the alg.
 //    to basically completely fail.
 //
+//    Mark C. Miller, Sun Jul  4 16:19:34 PDT 2021
+//    Simplified logic for NASTRAN exponential format
 // ****************************************************************************
-
-static float Getf(const char *s)
+static double GetVal(const char *s)
 {
-    char *ends;
-    double val = 0.0;
+    double dval;
+    char *endptr = 0;
+    size_t i;
+
+    // handle a string with no numbers
+    i = strcspn(s, "0123456789");
+    if (s[i] == '\0') return 0;
 
-    // Check for one of these funky 'NASTRAN exponential format' strings.
+    // Check for the funky 'NASTRAN exponential format'.
     // This is where a value like '1.2345e-5' is actually represented in the
     // file as '1.2345-5' with the 'e' character missing. It is awkward but
     // apparently a NASTRAN standard. I guess the rationale is that given
     // an 8 character field width limit, removing the 'e' character gives them
     // one additional digit of precision. This logic is basically looking for
     // the condition of encountering a sign character, '-' or '+', AFTER having
     // seen characters that could represent part of a number. In such a case,
-    // it MUST be the sign of the exponent.
-    const char *p = s;
-    char tmps[32];
-    char *q = tmps;
-    bool haveSeenNumChars = false;
-    while (!haveSeenNumChars || (*p != '-' && *p != '+' && *p != '\0'))
-    {
-        if (('0' <= *p && *p <= '9') || *p == '.' || *p == '+' || *p == '-')
-            haveSeenNumChars = true;
-        *q++ = *p++;
-    }
-    if (haveSeenNumChars && (*p == '-' || *p == '+'))
-    {
-        *q++ = 'e';
-        while (*p != '\0')
-            *q++ = *p++;
-        *q++ = '\0';
-        errno = 0;
-        val = strtod(tmps, &ends);
-    }
-    else
-    {
-        errno = 0;
-        val = strtod(s, &ends);
-    }
+    // it MUST be the sign of an exponent.
+    std::string t(s);
+    for (i = t.size(); t[i]!='-' && t[i]!='+' && t[i]!='\0'; i--);
+    if (i>0 && (t[i]=='-' || t[i]=='+') && t[i-1]!='e' && t[i-1]!='E')
+        t.insert(i, "E");
+
+    errno = 0;
+    dval = strtod(t.c_str(), &endptr);
 
-    if (errno != 0)
+    // Handle cases where conversion failed for one reason or another
+    if ((dval == 0 && endptr == t.c_str()) || errno)
     {
-        char msg[512];
-        snprintf(msg, sizeof(msg),
-            "Error \"%s\" at word \"%32s\"\n", strerror(errno), s);
-        if (!avtCallback::IssueWarning(msg))
-            cerr << msg << endl;
-        return 0.0;
+        static int nwarnings = 0;
```

Comment:
Do you really want this to be a static int? I believe this will only be initialized once during the execution of the program, so, for example, if this is called 5 times and each time it encounters the error 8 times, then it will only display this message when called the first time. This is ok if that is the intended behavior, just asking the question.

Reply:
Right. It would be better to occur once per database. I could I guess cache the database name here and then reset the counter whenever a different database is seen than the cached name. If the output is NOT constrained in some way, you can wind up getting problem-sized output on `cerr` for a database that is, for one reason or another, not being correctly processed...which is what I was observing with our test suite data. Another option...which might be better...is to remove all the warning logic all together.

Reply:
I think you solution is ok. Just thought I'd ask the question.